### PR TITLE
Connection availability tracker

### PR DESCRIPTION
**Pull request by gavin-norman-sociomantic**
Addresses #174.

### CODE REVIEW

**Comment by gavin-norman-sociomantic** [comment]
Rough, initial implementation of #174. As potential users, maybe could @vasileios-kyrillidis-sociomantic @matthias-wende-sociomantic have a look?

File: src/dhtproto/client/legacy/internal/helper/NodeAvailability.d [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+        public void error ( )
+        {
+            this.last_error_timestamp = time(null);
+        }
+
+        /// TODO
+        public bool available ( )
+        {
+            const delay_ms = 3_000; // TODO: configurable
+            auto diff = time(null) - this.last_error_timestamp;
```
Comment (nemanja-boric-sociomantic) [inline src/dhtproto/client/legacy/internal/helper/NodeAvailability.d]: We should use monotonic clock here which is not prone to adjustments.

File: src/dhtproto/client/legacy/internal/helper/NodeAvailability.d [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+    /// TODO
+    private struct NodeErrorTracker
+    {
+        /// TODO
+        private time_t last_error_timestamp;
+
+        /// TODO
+        public void error ( )
+        {
+            this.last_error_timestamp = time(null);
```
Comment (nemanja-boric-sociomantic) [inline src/dhtproto/client/legacy/internal/helper/NodeAvailability.d]: ditto monotonic clock.

**Review by vasileios-kyrillidis-sociomantic — COMMENTED** [review]
Thankfully Nemanja already spotted something I wouldn't :D
Apart from that it looks good to me! 

A bit offtopic: In general I prefer builder type objects to structs when they contain container instances (hash map in this case), but this is a perfectly acceptable solution so it LGTM!

**Comment by gavin-norman-sociomantic** [comment]
>A bit offtopic: In general I prefer builder type objects to structs when they contain container instances (hash map in this case), but this is a perfectly acceptable solution so it LGTM!

What would a "builder type object" look like in this case?

**Comment by gavin-norman-sociomantic** [comment]
Updated with monotonic clock.

**Comment by gavin-norman-sociomantic** [comment]
@vasileios-kyrillidis-sociomantic also note that, as I suspected, it's simpler to track when an error happened, rather than when an error _didn't_ happen (like in your older implementation).

File: src/dhtproto/client/legacy/internal/helper/NodeAvailability.d [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+        {
+            this.last_error_timestamp = time();
+        }
+
+        /// TODO
+        public bool available ( )
+        {
+            const delay_ms = 3_000; // TODO: configurable
+            auto diff = time() - this.last_error_timestamp;
+            return diff >= delay_ms;
```
Comment (nemanja-boric-sociomantic) [inline src/dhtproto/client/legacy/internal/helper/NodeAvailability.d]: You're comparing seconds with milliseconds here.

I suggest you change `time` to return the time in milliseconds:

`return 1000*ts.tv_sec + ts.tv_nsec / 1_000_000`.

Comment (gavin-norman-sociomantic) [reply src/dhtproto/client/legacy/internal/helper/NodeAvailability.d]: Haha yes. I've not tested this yet ;)

**Comment by gavin-norman-sociomantic** [comment]
Updated with doc and (small) unittest.

@vasileios-kyrillidis-sociomantic would you be able to test this in your app? Trying it out in a real world app would be the best way of checking it.

**Comment by vasileios-kyrillidis-sociomantic** [comment]
Sure I ll test this hopefully tomorrow.

**Comment by gavin-norman-sociomantic** [comment]
Thanks! Blocking until tested.

**Comment by vasileios-kyrillidis-sociomantic** [comment]
> What would a "builder type object" look like in this case?

A class with setters. Pretty much the same, just with `.setNodes()` method instead.

**Comment by gavin-norman-sociomantic** [comment]
>A class with setters. Pretty much the same, just with .setNodes() method instead.

Mm, you mean instead of passing the nodes registry to the `available` method? https://github.com/sociomantic-tsunami/dhtproto/pull/184/files#diff-274c306349464b80c856d501990f38e7R78

**Comment by vasileios-kyrillidis-sociomantic** [comment]
Yes, I was thinking something like you set the nodes at some point in the beginning and then just use the availability methods. But as said, it's not a huge difference.

**Comment by gavin-norman-sociomantic** [comment]
I see. Yeah, it could be a slightly more convenient API. I think it doesn't make much difference, though, as wherever you use the connection tracker you by definition have access to the DHT client instance, which contains the nodes registry.

**Comment by vasileios-kyrillidis-sociomantic** [comment]
Sorry ended up not looking into this today. I promise I will on Monday :)

**Comment by vasileios-kyrillidis-sociomantic** [comment]
I have just finished testing this on a local DHT with 2 DHT nodes and it seems to work fine. 

Only thing I realized is that I would like to have an `error` method with a signature like the equivalent available method `public void error (hash_t key, IDhtNodeRegistryInfo registry)`.

At least in Booty, we mark a node to be "down" when a DHT Put fails, meaning that I have the record key at hand and would be nice to directly "query" the node item internally in NoveAvailability (as done in the available method).

**Comment by gavin-norman-sociomantic** [comment]
>At least in Booty, we mark a node to be "down" when a DHT Put fails, meaning that I have the record key at hand and would be nice to directly "query" the node item internally in NoveAvailability (as done in the available method).

Can't you use the `NodeItem` struct in the notification? https://github.com/sociomantic-tsunami/swarm/blob/v5.x.x/src/swarm/client/request/notifier/IRequestNotification.d#L134

**Comment by vasileios-kyrillidis-sociomantic** [comment]
Of course it can be done. It's just for 2 minor reasons I wrote the above:
1) It's just that in the current logic we time a request out after a small amount of time, meaning outside of the notifier. I would have to store a reference to the node item and use it outside (so no real problem here). 
2) I would expect the API to be similar for both "error" and "available"

**Comment by gavin-norman-sociomantic** [comment]
>It's just that in the current logic we time a request out after a small amount of time, meaning outside of the notifier.

Hm, can you explain this a bit more? Doesn't a request timeout also call the notifier? (Perhaps I'm forgetting some aspect of the legacy DHT client.)

>I would expect the API to be similar for both "error" and "available"

They were intended to be used from completely different contexts (`available` before assigning a request and `error` from inside a request's notifier), hence the different signatures.

**Comment by vasileios-kyrillidis-sociomantic** [comment]
> Doesn't a request timeout also call the notifier? (Perhaps I'm forgetting some aspect of the legacy DHT client.)

No not really, the code looks like this: 
```
    StopWatch stop_watch;
    stop_watch.start();
    while (!assignPutRequest(put_request)) // Notifier is in the assignPutRequest
    {
        // Check if we should time this request out
        // NOTE: It's ok to lose some precision here by casting to uint
        uint elapsed_millis = cast(uint)(stop_watch.microsec / 1000);
        if (elapsed_millis >= timeout_wait_millis)
        {
            auto node_item = getResponsibleDhtNodeItem(key);
            node_availability.error(node_item);
            return false;
        }

        // Otherwise wait for a bit and then retry
        wait(dht_config.retry_wait_millis);
    }
```

Edit: Basically we manually track the timeouts.

**Comment by gavin-norman-sociomantic** [comment]
That's... extremely unusual usage! Why are you doing it that way? Why not use the DHT client's timeout facilities, if that's what you need?

I don't feel it's a good idea to adapt the API to handle such unusual usage, unless you can explain a good reason why you have to do it that way.

**Comment by gavin-norman-sociomantic** [comment]
Verbal discussion: we decided to adapt the app to use the availability tracker as intended (removing its own tracking of retry times). @vasileios-kyrillidis-sociomantic approves this PR :)

**Comment by gavin-norman-sociomantic** [comment]
Unblocked following testing.

**Review by vasileios-kyrillidis-sociomantic — APPROVED** [review]

**Review by matthias-wende-sociomantic — APPROVED** [review]
:+1:

**Pull request merged**